src/core/net/dns_client.cpp

```diff
@@ -1109,6 +1183,112 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP
+void Client::HandleTcpTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
+}
+
+void Client::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    LogDebg("Dns tcp connection established.\n");
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpEstablished(aEndpoint);
+}
+
+void Client::HandleTcpSendDoneCallback(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aData);
+}
+
+void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                       size_t         aBytesAvailable,
+                                       bool           aEndOfStream,
+                                       size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message *             message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+
+    if (aEndOfStream)
+    {
+        // Already received end of stream timer isn't necessary cleanup is done in HandleTcpDisconnected
+        mTcpTimer.Stop();
+    }
+
+    VerifyOrExit(aBytesAvailable != 0);
+    mEndpoint.ReceiveByReference(data);
+    message = mSocket.NewMessage(0);
+
+    for (; data != nullptr; data = data->mNext)
+    {
+        totalReceived += data->mLength;
+        message->AppendBytes(data->mData + tcpQueryLenSize, data->mLength - tcpQueryLenSize);
```

Comment:
Just checking.. do we need to add `tcpQueryLenSize` for each `data` or just the first one?

Reply:
well looking at the current implementation I have just one buffer so iterating over them is overkill. So I changed this `for` to `if`.